### PR TITLE
Fourier

### PR DESCRIPTION
**Pull request by fsxbhyy**
- Add Green2DLR
- Add Fourier operation for Green2DLR
- Add test of Fourier operation
-Green2 is still wip, please check Green2DLR first

### CODE REVIEW

File: src/green/Green.jl [hunk 1]
```diff
@@ -38,25 +47,173 @@ mutable struct Green2{T<:AbstractFloat,TGT,SGT,CT}
...
+    - 'spaceGrid': Coordinate or momentum grid
+    - 'error': The error of noisy Green's function
+    - 'dlrGrid': In-built Discrete Lehmann Representation
+    """
+    function Green2DLR{T}(timeType,spaceType,isFermi, β, Euv, rtol,spaceGrid::SGT; color::CT=nothing, timeGrid::TGT=nothing,timeSymmetry=:none, spaceSymmetry=nothing,error=nothing)where{T<:Number,TGT,SGT,CT}
+        @assert timeType == :freq ||timeType == :tau ||timeType == :dlr
+        @assert spaceType == :mom ||timeType == :spa
+        @assert timeSymmetry == :ph ||timeSymmetry == :pha ||timeSymmetry == :none
+        dlrGrid = DLRGrid(Euv, β, rtol, isFermi, timeSymmetry)
+        ct = CT
```
Comment (kunyuan) [inline src/green/Green.jl:128]: move this liine to 140-142, where ct is redefined

File: src/green/Green.jl [hunk 2]
```diff
@@ -38,25 +47,173 @@ mutable struct Green2{T<:AbstractFloat,TGT,SGT,CT}
...
+    - 'β': Inverse temperature
+    - 'Euv': the UV energy scale of the spectral density
+    - 'rtol': tolerance absolute error
+    - 'color': Indices of species of Green's function (such as different spin values). Default: One element array [1]
+    - 'timeGrid': Time or Frequency grid. Default: DLR grid in timeType (frequency, tau or dlr) 
+    - 'spaceGrid': Coordinate or momentum grid
+    - 'error': The error of noisy Green's function
+    - 'dlrGrid': In-built Discrete Lehmann Representation
+    """
+    function Green2DLR{T}(timeType,spaceType,isFermi, β, Euv, rtol,spaceGrid::SGT; color::CT=nothing, timeGrid::TGT=nothing,timeSymmetry=:none, spaceSymmetry=nothing,error=nothing)where{T<:Number,TGT,SGT,CT}
```
Comment (kunyuan) [inline src/green/Green.jl:123]: reorder the parameters, so that the related parameters stay close with each other

File: src/green/Green.jl [hunk 3]
```diff
@@ -38,25 +47,173 @@ mutable struct Green2{T<:AbstractFloat,TGT,SGT,CT}
...
+            aimGrid = green.dlrGrid.n
+        elseif(aimType == :tau)
+            aimGrid = green.dlrGrid.τ
+        elseif(aimType == :dlr)
+            aimGrid = green.dlrGrid.ω
+        end
+    end
+    if(green.timeType==:freq)
+        if(aimType == :tau)
+            data_mid = matfreq2dlr(green.dlrGrid, green.dynamic, green.timeGrid; green.error,axis=4)
```
Comment (kunyuan) [inline src/green/Green.jl:191]: use matfreq2tau instead of matfreq2dlr, then dlr2tau

File: src/green/Green.jl [hunk 4]
```diff
@@ -38,25 +47,173 @@ mutable struct Green2{T<:AbstractFloat,TGT,SGT,CT}
...
+    function TimeFourier(green::Green2DLR, aimType::Symbol; aimGrid = nothing)
+    Convert Green's function to different time representation (frequency, tau , or dlr).
+    #Arguements
+    - 'green': Original Green's function
+    - 'aimType': Time representation of outcome Green's function
+    - 'aimGrid': Grid of outcome Green's function. Default: DLR grid of aimType
+"""
+
+
+function TimeFourier(green::Green2DLR, aimType::Symbol; aimGrid = nothing)
```
Comment (kunyuan) [inline src/green/Green.jl:179]: am --> target ?

File: src/green/Green.jl [hunk 5]
```diff
@@ -38,25 +47,173 @@ mutable struct Green2{T<:AbstractFloat,TGT,SGT,CT}
...
+        end
+    elseif(green.timeType == :dlr)
+        if(aimType == :freq)
+            data = dlr2matfreq(green.dlrGrid, green.dynamic, aimGrid; axis=4)
+        elseif(aimType == :tau)
+            data = dlr2tau(green.dlrGrid, green.dynamic, aimGrid; axis=4)
+        end
+    end
+
+    green_new = Green2DLR{typeof(data[1,1,1,1])}(aimType,green.spaceType, green.isFermi, green.β, green.Euv, green.rtol, green.spaceGrid; color = green.color, timeGrid = aimGrid, timeSymmetry = green.timeSymmetry, spaceSymmetry = green.spaceSymmetry, error = green.error)
```
Comment (kunyuan) [inline src/green/Green.jl:211]: typeof(data[1,1,1,1]) --> eltype(data)

File: src/green/Green.jl [hunk 6]
```diff
@@ -38,25 +47,173 @@ mutable struct Green2{T<:AbstractFloat,TGT,SGT,CT}
...
+    green_new.dynamic = data
+    green_new.instant = green.instant
```
Comment (kunyuan) [inline src/green/Green.jl:213]: instant will change by a factor \beta in fourier transform

File: src/green/Green.jl [hunk 7]
```diff
@@ -38,25 +47,173 @@ mutable struct Green2{T<:AbstractFloat,TGT,SGT,CT}
...
+        elseif(aimType == :dlr)
+            aimGrid = green.dlrGrid.ω
+        end
+    end
+    if(green.timeType==:freq)
+        if(aimType == :tau)
+            data_mid = matfreq2dlr(green.dlrGrid, green.dynamic, green.timeGrid; green.error,axis=4)
+            data = dlr2tau(green.dlrGrid, data_mid, aimGrid; axis=4)
+        elseif(aimType == :dlr)
+            data = matfreq2dlr(green.dlrGrid, green.dynamic, green.timeGrid; green.error,axis=4)
```
Comment (kunyuan) [inline src/green/Green.jl:194]: rename data --> dynamic

File: src/green/Green.jl [hunk 8]
```diff
@@ -38,25 +47,177 @@ mutable struct Green2{T<:AbstractFloat,TGT,SGT,CT}
...
+                timeGrid = dlrGrid.ω
+            end
+            tgt = typeof(timeGrid)
+        end
+        ct = CT
+        if(color == nothing)
+            color = [1]
+            ct = typeof(color)
+        end
+        instant = zeros(T, (length(color),length(color),length(spaceGrid)))
```
Comment (kunyuan) [inline src/green/Green.jl:144]: two error: instantError, dynamicError; they have the same type as the array instant, dynamic

File: src/green/Green.jl [hunk 9]
```diff
@@ -38,25 +47,177 @@ mutable struct Green2{T<:AbstractFloat,TGT,SGT,CT}
...
+"""
+    function TimeFourier(green::Green2DLR, aimType::Symbol; aimGrid = nothing)
+    Convert Green's function to different time representation (frequency, tau , or dlr).
+    The fourier convention of instant part is:  G_ins(ωn) = G_ins(ω) = β*G_ins(τ) = G_ins
+    #Arguements
+    - 'green': Original Green's function
+    - 'aimType': Time representation of outcome Green's function
+    - 'aimGrid': Grid of outcome Green's function. Default: DLR grid of aimType
+"""
+
```
Comment (iintSjds) [inline src/green/Green.jl:176]: comments for doc should stick to the definition. Remove blank line 176&177

File: src/green/Green.jl [hunk 10]
```diff
@@ -38,25 +47,194 @@ mutable struct Green2{T<:AbstractFloat,TGT,SGT,CT}
...
+
+function TimeFourier(green::Green2DLR, targetType::Symbol; targetGrid = nothing)
```
Comment (kunyuan) [inline src/green/Green.jl:182]: name convention!!! The first letter should be lower case

File: src/green/Green.jl [hunk 11]
```diff
@@ -38,25 +47,194 @@ mutable struct Green2{T<:AbstractFloat,TGT,SGT,CT}
...
+    # end
 end
 
+"""
+    function TimeFourier(green::Green2DLR, aimType::Symbol; aimGrid = nothing)
+    Convert Green's function to different time representation (frequency, tau , or dlr).
+    The fourier convention of instant part is:  G_ins(ωn) = G_ins(ω) = β*G_ins(τ) = G_ins
+    #Arguements
+    - 'green': Original Green's function
+    - 'aimType': Time representation of outcome Green's function
```
Comment (kunyuan) [inline src/green/Green.jl:177]: aimType --> targetType

**Pull request merged**